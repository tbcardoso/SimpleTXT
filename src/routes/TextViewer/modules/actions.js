/*
 * action types
 */

export const SELECT_FILE_SUCCESS = 'SELECT_FILE_SUCCESS'
export const LOAD_FILE_SUCCESS = 'LOAD_FILE_SUCCESS'



/*
 * action creators
 */

export function selectFileSuccess(filePath) {
  return { type: SELECT_FILE_SUCCESS, filePath }
}

export function loadFileSuccess(filePath, fileContent) {
  return { type: LOAD_FILE_SUCCESS, filePath, fileContent }
}