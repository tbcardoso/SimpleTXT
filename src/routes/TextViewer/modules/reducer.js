import { combineReducers } from 'redux';
import * as Actions from './actions';
import about from '../../../assets/about';

const selectedFileInitialState = {
	filePath: '',
	fileContent: '',
	loadingFilePath: '',
	error: '',
};

function selectedFile(state = selectedFileInitialState, action) {
	switch (action.type) {
		case Actions.SELECT_FILE_SUCCESS:
			return { ...state, loadingFilePath: action.filePath };
		case Actions.LOAD_FILE_SUCCESS:
			return { ...state, loadingFilePath: '', filePath: action.filePath, fileContent: action.fileContent, error: '' };
		case Actions.LOAD_FILE_ERROR:
			return { ...state, loadingFilePath: '', error: action.error };
		case Actions.ABOUT_SCREEN_SELECTED:
			return { ...state, loadingFilePath: '', filePath: 'About', fileContent: about, error: '' };
		default:
			return state;
	}
}

export default combineReducers({
	selectedFile,
});
