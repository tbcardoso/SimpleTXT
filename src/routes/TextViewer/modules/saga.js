import { call, put, takeLatest } from 'redux-saga/effects';
import * as Actions from './actions';
import RNFetchBlob from 'react-native-fetch-blob';
import { ToastAndroid } from 'react-native';

function * selectedFileSaga(action) {
	console.log('Selected file Saga: ', action);

	try {
		const data = yield call(RNFetchBlob.fs.readFile, action.filePath, 'utf8');
		//console.log('File data: ', data);
		yield put(Actions.loadFileSuccess(action.filePath, data));
	} catch (e) {
		console.log('Load File Error: ', e);
		ToastAndroid.show(e.message, ToastAndroid.SHORT)
		yield put(Actions.loadFileError(action.filePath, e.message));
	}
}

export default function * saga() {
	yield takeLatest(Actions.SELECT_FILE_SUCCESS, selectedFileSaga);
}
