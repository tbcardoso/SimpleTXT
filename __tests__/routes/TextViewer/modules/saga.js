/* eslint-env jest */
// require('react-native-mock/mock');
import * as Actions from '../../../../src/routes/TextViewer/modules/actions';
import * as Sagas from '../../../../src/routes/TextViewer/modules/saga';
// import MockRealm from '../../../../__mocks__/realm';

jest.mock('react-native-fetch-blob');
jest.mock('realm');
// jest.mock('react-native');

describe('TextViewer sagas', () => {
	test('Saga loadStartupFileSaga', () => {
		// const mockRealm = new MockRealm();
		// console.log('mockRealm: ', mockRealm);
		// let files = mockRealm.objects();
		// console.log('files: ', files);
		// console.log('file.length: ', files.length);

		const saga = Sagas.loadStartupFileSaga(Actions.loadStartupFile());
		expect(saga.next().value).toMatchSnapshot();
		expect(saga.next().done).toBe(true);

		// return expectSaga(Sagas.loadStartupFileSaga, Actions.loadStartupFile())
		// 	.put(Actions.selectFileSuccess('path/to/file.txt'))
		// 	.run();
	});

	test('Saga selectedFileSaga', () => {
		const saga = Sagas.selectedFileSaga(Actions.selectFileSuccess('path/to/file.txt'));
		expect(saga.next().value).toMatchSnapshot();
		expect(saga.next('Lorem ipsum...').value).toMatchSnapshot();
		expect(saga.next().done).toBe(true);
	});

	test('Saga selectedFileSaga exception', () => {
		const saga = Sagas.selectedFileSaga(Actions.selectFileSuccess('path/to/file.txt'));
		expect(saga.next().value).toMatchSnapshot();
		expect(saga.throw({ message: 'Error' })).toMatchSnapshot();
		expect(saga.next().done).toBe(true);
	});

});
