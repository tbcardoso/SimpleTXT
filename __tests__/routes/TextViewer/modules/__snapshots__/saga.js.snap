// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextViewer sagas Saga loadStartupFileSaga 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "filePath": "path/to/file.txt",
      "type": "SELECT_FILE_SUCCESS",
    },
    "channel": null,
  },
}
`;

exports[`TextViewer sagas Saga selectedFileSaga 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "path/to/file.txt",
      "utf8",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`TextViewer sagas Saga selectedFileSaga 2`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "fileContent": "Lorem ipsum...",
      "filePath": "path/to/file.txt",
      "type": "LOAD_FILE_SUCCESS",
    },
    "channel": null,
  },
}
`;

exports[`TextViewer sagas Saga selectedFileSaga exception 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "path/to/file.txt",
      "utf8",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`TextViewer sagas Saga selectedFileSaga exception 2`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "PUT": Object {
      "action": Object {
        "error": "Error",
        "filePath": "path/to/file.txt",
        "type": "LOAD_FILE_ERROR",
      },
      "channel": null,
    },
  },
}
`;
